server/route.go

```diff
@@ -1920,23 +1988,35 @@ func (c *client) processRouteConnect(srv *Server, arg []byte, lang string) error
 	}
 
 	clusterName := srv.ClusterName()
+	isSusceptible := srv.isClusterNameSusceptibleToChanges()
 
 	// If we have a cluster name set, make sure it matches ours.
 	if proto.Cluster != clusterName {
-		shouldReject := true
 		// If we have a dynamic name we will do additional checks.
-		if srv.isClusterNameDynamic() {
-			if !proto.Dynamic || strings.Compare(clusterName, proto.Cluster) < 0 {
-				// We will take on their name since theirs is configured or higher then ours.
-				srv.setClusterName(proto.Cluster)
-				if !proto.Dynamic {
-					srv.getOpts().Cluster.Name = proto.Cluster
-				}
+		isDynamic := srv.isClusterNameDynamic()
+		remoteIsDynamic := proto.Dynamic
+
+		if isDynamic && !remoteIsDynamic {
+			// c.Noticef("Updating the explicit name from the CONNECT message: ", proto.Cluster)
+			// Use the explicit name and mark current node as not susceptible
+			// to dynamic cluster name changes.
+			srv.setClusterName(proto.Cluster, false)
+			srv.removeAllRoutesExcept(c)
+		} else if !isDynamic && remoteIsDynamic {
+			// Do nothing since the remote may use this node's cluster name
+			// when it is sent an INFO message. Also override the cluster name
+			// from this route with ours since should win.
+			c.mu.Lock()
+			c.route.clusterName = srv.info.Cluster
```

Comment:
technically you would need to capture the cluster name before getting the c.mu.Lock() because info.Cluster can change (during reload).